### PR TITLE
#693 and #660 and #698

### PR DESCRIPTION
**Pull request by rmandelb**
This branch includes the following updates:

- The ability to initialize an InterpolatedImage from a specific HDU of a fits file (#660).
- Some updates to the COSMOSCatalog and the config interface to allow pre-selection of galaxies being used for simulations, either using parametric fits or realistic galaxies (#693).
- Some updates to the catalogs themselves to enable that pre-selection to happen correctly and efficiently (which means that taking advantage of some updates requires re-downloading the catalogs; GalSim will identify whether that is needed, and the ```galsim_download_cosmos``` utility can be used for the downloading and unpacking).  Plus there's a new COSMOS catalog available, which isn't technically a GalSim update except that it is now the default one that GalSim will use unless the original one is specified, and there are some new kwargs that can be used to choose between the currently available catalogs.

Since both Mike and I worked on parts of this, it would be helpful if someone else like @jmeyers314 could take a look.  And I think @beckermr was interested in this, so perhaps he'd like to try it out?

### CODE REVIEW

**Comment by rmjarvis** [comment]
I also did #698 on this branch: deprecating the config 'Ring' object type and putting this functionality in a stamp type (again called 'Ring') instead.

**Comment by jmeyers314** [comment]
>it would be helpful if someone else like @jmeyers314 could take a look

Sure thing.

**Comment by jmeyers314** [comment]
I just started looking this over by cleaning out my GalSim directory with:

```
scons -c
rm -rf .scon*
```

and then compiling with plain `scons`.

For quite some time now (certainly nothing specific to this PR), this has been generating warnings on my laptop; all of which look like:

```
1 warning generated.
In file included from pysrc/CorrelatedNoise.cpp:28:
In file included from include/galsim/CorrelatedNoise.h:29:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/TMV_Sym.h:27:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/TMV.h:30:
In file included from /Users/josh/miniconda2/opt/lsst/tmv/include/tmv/TMV_Matrix.h:495:
/Users/josh/miniconda2/opt/lsst/tmv/include/tmv/TMV_Permutation.h:465:57: warning: dependent nested
      name specifier 'VectorView<T, A>::' for friend class declaration is not supported; turning off
      access control for 'Permutation' [-Wunsupported-friend]
        friend inline VectorView<T,A>& VectorView<T,A>::sort(
                                       ~~~~~~~~~~~~~~~~~^
```

**Comment by rmjarvis** [comment]
> For quite some time now (certainly nothing specific to this PR), this has been generating warnings on my laptop; all of which look like:

Yes.  Sorry.  I know about those.  clang added this warning recently, and I haven't gotten around to fixing TMV to avoid it yet.  Until I do, you'll need to compile with WARN=false unfortunately.

File: galsim/scene.py [hunk 1]
```diff
@@ -33,7 +33,8 @@
 
 class COSMOSCatalog(object):
     """
-    A class representing a sample of galaxies from the COSMOS sample with F814W<23.5.
+    A class representing a random subsample of galaxies from the COSMOS sample with F814W<25.2
```
Comment (jmeyers314) [inline galsim/scene.py]: Should probably indicate at which galsim version number the default catalog changes, i.e. the meaning of "previous" here.

File: galsim/scene.py [hunk 2]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+                            F814W<25.2 COSMOS catalog in $PREFIX/share/galsim.  It will raise an
+                            exception if the catalog is not there telling you to run
+                            galsim_download_cosmos.]
+    @param sample           A keyword argument that can be used to specify the sample to use, i.e.,
+                            "23.5" or "25.2".  At most one of `file_name` and `sample` should be
+                            specified.
+                            [default: None, which results in the same default as `file_name=None`.]
+    @param image_dir        Keyword that is only used for real galaxies, not parametric ones, to
+                            specify the directory of the image files.
+                            If a string containing no `/`, it is the relative path from the location
```
Comment (jmeyers314) [inline galsim/scene.py]: Is there a reason not to follow the normal *nix relative-vs-absolute path rule here?  I.e., relative paths can't *start* with '/', but could contain '/' at later positions?  I see there's even an os.path.isabs() utility for this.

File: galsim/scene.py [hunk 3]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+                            of the catalog file to the directory containing the galaxy/PDF images.
+                            If a path (a string containing `/`), it is the full path to the
+                            directory containing the galaxy/PDF images. [default: None]
+    @param dir              The directory of catalog file. [default: None]
+    @param preload          Keyword that is only used for real galaxies, not parametric ones, to
+                            choose whether to preload the header information.  If `preload=True`,
+                            the bulk of the I/O time is in the constructor.  If `preload=False`,
+                            there is approximately the same total I/O time (assuming you eventually
+                            use most of the image files referenced in the catalog), but it is spread
+                            over the various calls to getGal() and getPSF().  [default: False]
```
Comment (jmeyers314) [inline galsim/scene.py]: Should `getGal()` and `getPSF()` just be `makeGalaxy()` here?

File: galsim/scene.py [hunk 4]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+    @param noise_dir        Keyword that is only used for real galaxies, not parametric ones.
```
Comment (jmeyers314) [inline galsim/scene.py]: Does `noise_dir` need any relative-vs-absolute path guidance like `image_dir`?

File: galsim/scene.py [hunk 5]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+                            The directory of the noise files if different from the directory of the 
+                            image files.  [default: image_dir]
+    @param use_real         Enable the use of realistic galaxies?  [default: True]
```
Comment (jmeyers314) [inline galsim/scene.py]: Is there a speed or RAM benefit to having `use_real=False` when making parametric galaxies?  If so, then should add this to the documentation.  If not, then why have this kwarg?

File: galsim/scene.py [hunk 6]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+                            If this parameter is False, then `makeGalaxy(gal_type='real')` will
+                            not be allowed.
+    @param exclusion_level  Level of cuts to make on the galaxies based on the quality of postage
+                            stamp definition and/or parametric fit quality.  Options:
+                            "none": No cuts.
```
Comment (jmeyers314) [inline galsim/scene.py]: I assume there were some cuts in place when making the catalogs.  So should this be "No additional cuts", maybe?

File: galsim/scene.py [hunk 7]
```diff
@@ -86,46 +92,58 @@ class COSMOSCatalog(object):
...
+                            "bad_ps": Apply cuts to eliminate galaxies that have failures in postage
```
Comment (jmeyers314) [inline galsim/scene.py]: I see `bad_ps` and immediately think that something is wrong with a power spectrum somewhere.  Suggest `bad_stamp` as an alternative.

File: galsim/scene.py [hunk 8]
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
         self.orig_index = np.arange(len(self.param_cat))
         mask = np.ones(len(self.orig_index), dtype=bool)
 
-        # If requested, select galaxies based on existence of a usable fit.
-        if exclude_fail:
+        # If requested, select galaxies in a way that excludes suspect postage stamps (e.g., with
+        # deblending issues), suspect parametric model fits, or both of the above plus marginal
+        # ones.
+        if exclusion_level == 'bad_ps' or exclusion_level == 'marginal':
```
Comment (jmeyers314) [inline galsim/scene.py]: Looks like this is the same if statement as line 239.  Could these be streamlined?

File: galsim/scene.py [hunk 9]
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
...
+                sn_limit = 12.0
+            if self.selection_cat is not None:
+                div_val = self.selection_cat['peak_image_pixel_count']
+                div_val[div_val == 0.] = 1.e-5
+                mask &= ( (self.selection_cat['sn_ellip_gauss'] >= sn_limit) &
+                          ((self.selection_cat['min_mask_dist_pixels'] > 11.0) |
+                           (self.selection_cat['average_mask_adjacent_pixel_count'] / \
+                               div_val < cut_ratio)) )
+
+        if exclusion_level == 'bad_fits' or exclusion_level == 'marginal':
```
Comment (jmeyers314) [inline galsim/scene.py]: Personally, I like the earlier style: `if exclusion_level in ['bad_fits', 'marginal']`.

File: galsim/scene.py [hunk 10]
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
...
+                      (sersicfit_status < 5)) &
+                      ((bulgefit_status > 0) &
+                      (bulgefit_status < 5)) )
+        
+            # Some fit parameters can indicate a likely sky subtraction error
             hlr = self.param_cat['sersicfit'][:,1]
             n = self.param_cat['sersicfit'][:,2]
-            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) ) 
-            # May add more cuts here if we discover other kinds of problematic objects.
+            mask &= ( (n < 5) | (hlr < 1./cosmos_pix_scale) )
```
Comment (jmeyers314) [inline galsim/scene.py]: I think this is saying that the HLR should be smaller than 1 arcsec.  This would be clearer with something like
```
hlr_pix = self.param_cat['sersicfit'][:, 1]
max_hlr_arcsec = 1.0
mask &= ((n < 5) | (hlr_pix < max_hlr_arcsec/cosmos_pix_scale))
```

Comment (jmeyers314) [inline galsim/scene.py]: Also, I see down below that there's also a factor of `np.sqrt(q)` in the formula for hlr in arcsec.  Maybe a comment would be good here explaining why that's important down below but not here.

File: galsim/scene.py [hunk 11]
```diff
@@ -194,47 +271,104 @@ def __init__(self, file_name=None, image_dir=None, dir=None, preload=False, nois
...
+                hlr_pix = sparams[:,1]
+                n = sparams[:,2]
+                q = sparams[:,3]
+                hlr = cosmos_pix_scale*hlr_pix*np.sqrt(q)
+                if min_flux > 0. or max_flux > 0.:
+                    flux_hlr = sparams[:,0]
+                    # The prefactor for n=4 is 3.607.  For n=1, it is 1.901.
+                    # It's not linear in these values, but for the sake of efficiency and the 
+                    # ability to work on the whole array at once, just linearly interpolate.
+                    # Hopefully, this can be improved as part of issue #693.  Maybe by storing the
```
Comment (jmeyers314) [inline galsim/scene.py]: I assume this comment is outdated?

File: galsim/scene.py [hunk 12]
```diff
@@ -401,7 +549,7 @@ def _makeParametric(self, indices, chromatic, sersic_prec, gsparams):
 
         gal_list = []
         for index in indices:
-            record = self.param_cat[self.orig_index[index]]
+            record = self.getParametricRecord(index)
```
Comment (jmeyers314) [inline galsim/scene.py]: Looks like I've been commenting on a fair amount of old code that just happened to show up in the imperfectly processed github diff.  Hope that's not too annoying.  I actually have a comment for some old code on line 751 in `getParametricRecord`, which isn't imperfectly diffed, so github won't let me comment on that line, so I'll place it here instead.  I was happy to see a generator expression:

```
record_dict = dict( ((k,record[k]) for k in record.dtype.names) )
```

I just wanted to point out that a generator expression passed as a single argument to a function doesn't need its own parens.  I.e., the following works and, at least to me, looks a bit cleaner:

```
record_dict = dict((k,record[k]) for k in record.dtype.names)
```

File: galsim/scene.py (hunk 1 above)
Comment (rmandelb) [inline galsim/scene.py]: Indeed.  Done!

File: galsim/scene.py (hunk 3 above)
Comment (rmandelb) [inline galsim/scene.py]: Yes they should.  That was clearly copied from the RealGalaxyCatalog description.  Fixed.

File: galsim/scene.py (hunk 9 above)
Comment (rmandelb) [inline galsim/scene.py]: Fixed.

File: galsim/scene.py (hunk 11 above)
Comment (rmandelb) [inline galsim/scene.py]: Yes, fixed.

File: galsim/scene.py (hunk 10 above)
Comment (rmandelb) [inline galsim/scene.py]: Both the sqrt(q) issue and the nature of this cut itself are clarified in https://github.com/GalSim-developers/GalSim/commit/e7761cdd750d921e786282aae1349ee6942811f5

File: galsim/scene.py (hunk 8 above)
Comment (rmandelb) [inline galsim/scene.py]: Good catch.  Done: https://github.com/GalSim-developers/GalSim/commit/7e3b9aa4889cd4d976daf7af47150bd4662abf3f

File: galsim/scene.py (hunk 7 above)
Comment (rmandelb) [inline galsim/scene.py]: Fair enough.  Done.

File: galsim/scene.py (hunk 2 above)
Comment (rmandelb) [inline galsim/scene.py]: Hmmm, interesting question.   @rmjarvis , what are your thoughts?

File: galsim/scene.py (hunk 6 above)
Comment (rmandelb) [inline galsim/scene.py]: Well - those cuts were really minimal because we were trying to get as close as we could to a flux-limited sample.  But I suppose technically we should mention them and point to their description.  Done.

File: galsim/scene.py (hunk 2 above)
Comment (rmjarvis) [inline galsim/scene.py]: I've always found the way these dir kwargs are parsed a bit confusing, to be honest.  I'd actually rather have just a single dir for everything and let the user use symlinks if they want to put things in different directories than the standard layout.  Seems to me that would be less confusing than having 3 different dir kwargs here.  Then there wouldn't be any need to distinguish between absolute and relative paths.

Comment (rmandelb) [inline galsim/scene.py]: I'm fine with that in principle.  But it might be tricky to do this in a way that doesn't break people's old code when they go to v1.4.  If you have ideas for how to cope with that, I would be fine with updating both COSMOSCatalog and RealGalaxyCatalog to this new model.

File: galsim/scene.py (hunk 12 above)
Comment (rmandelb) [inline galsim/scene.py]: Done.

File: galsim/scene.py (hunk 5 above)
Comment (rmandelb) [inline galsim/scene.py]: Well - in the very original version of the code I think there may have been more benefit because it was structured differently.  Now, there is a quite small benefit in terms of RAM and speed (don't have to read in the real catalog) and a possible logistical benefit, like if you never want to use realistic galaxies and deleted the real catalog from your system, you could still use COSMOSCatalog with ```use_real=False``` and not have it complain that it couldn't find the catalog.  I just added a note on this.

File: galsim/scene.py (hunk 2 above)
Comment (rmjarvis) [inline galsim/scene.py]: It would mean deprecating the old functionality, rather than removing it.  For now, probably just leave the code the same, but give deprecation warnings if either `image_dir` or `noise_dir` is not None.  I don't think there is any change required for the case where `dir` is given, but not `image_dir` or `noise_dir`.

File: galsim/scene.py [hunk 13]
```diff
@@ -350,13 +504,24 @@ def makeGalaxy(self, index=None, gal_type=None, chromatic=False, noise_pad_size=
...
+                    'You seem to have an old version of the COSMOS parameter file.\n'+
+                    'Please run `galsim_download_cosmos` to re-download the COSMOS catalog\n' +
+                    'and take advantage of pre-computation of many quantities..')
+
             gal_list = self._makeParametric(indices, chromatic, sersic_prec, gsparams)
 
-        # If deep, rescale the size and flux
+        # If trying to use the 23.5 sample and "fake" a deep sample, rescale the size and flux as
+        # suggested in the GREAT3 handbook.
         if deep:
```
Comment (jmeyers314) [inline galsim/scene.py]: Should the `deep` kwarg be disabled if using the 25.2 catalog?  (Same question for `_makeSingleGalaxy` down below).

Comment (rmandelb) [inline galsim/scene.py]: Yes.  Done.

File: galsim/scene.py (hunk 2 above)
Comment (rmandelb) [inline galsim/scene.py]: OK.  I will take a stab at it this weekend; will post when it�s ready to go over.

File: examples/great3/cgc.yaml [hunk 14]
```diff
@@ -16,9 +16,11 @@
 #    and/or other materials provided with the distribution.
 
 # The Great3 control-ground-constant branch configuration file
-# Note: This configuration will not produce images _identical_ to the ones used for Great3, 
+# Note: This configuration will not produce images _identical_ to the ones used for Great3,
 # since the order of random number generation is different.  However, it is intended to
 # produce an _equivalent_ simulation as the one produced for Great3.
+modules:
```
Comment (jmeyers314) [inline examples/great3/cgc.yaml]: I know that this example isn't really part of the tutorial series, but I think it would be nice for some more documentation on the `modules` top-level config type here.  (I think the only other example of this is `examples/des/draw_psf.yaml`.)  Or alternatively, maybe a new tutorial that more simply shows off the `modules` config type (or augment an existing tutorial?).

File: examples/great3/cgc_faint.yaml [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+
+# An alterate version of the Great3 control-ground-constant branch configuration file, 
+# using an I<25.2 sample (instead of faking a deeper sample using I<23.5 images).
+
+# Start with the cgc.yaml file, which we'll modify below.
+template: cgc.yaml
```
Comment (jmeyers314) [inline examples/great3/cgc_faint.yaml]: This top-level config type is also really cool.  We should advertise it in the tutorial series!

File: examples/great3/cgc.yaml (hunk 14 above)
Comment (rmjarvis) [inline examples/great3/cgc.yaml]: The DES examples are a more complete set of examples that use the modules field in various ways.  cf. the PR for #654.  I've called both the DES and the Great3 examples "Advanced Simulations" [here](https://github.com/GalSim-developers/GalSim/wiki/Tutorials) and mention the top-level modules field as a significant feature of both.

I'm not averse to including the modules in the main set of numbered demos.  I think the natural place will likely be demo13.  But that will have to wait until the chromatic stuff is folding into the config framework, which I haven't done yet.

File: examples/great3/cgc.yaml [hunk 16]
```diff
@@ -122,16 +132,32 @@ gal:
...
-    # lower S/N.  To compensate, they added less noise in those cases.  However, the
-    # exact formula they used is still a bit of a mystery to me.
-    # They did this as a LookupTable, but the formula below is pretty close.
-    fmax_var: '$0.004 * math.exp(2.9 * (1.1-@psf.items.0.fwhm) * (0.95-@psf.items.0.fwhm))'
+    # different seeing sizes.  When the seeing is larger, a given flux galaxy will have
+    # lower S/N.  To compensate, they added less noise in those cases.  The formula that was used
+    # was based on running experiments with different seeing values, and checking how the S/N varied
+    # at fixed noise variance.  The results of those experiments were used to create a LookupTable
+    # in the GREAT3 scripts, but the formula below is pretty close.
+    fmax_var: '$0.004 * math.exp(2.9 * (1.1-psf_fwhm) * (0.95-psf_fwhm))'
```
Comment (jmeyers314) [inline examples/great3/cgc.yaml]: Are the '$' and '@' Eval shorthands documented somewhere?

Comment (rmjarvis) [inline examples/great3/cgc.yaml]: It's not live yet.  But yes.  Here: https://github.com/GalSim-developers/GalSim/wiki/Config-Values#shorthand-notation

**Comment by jmeyers314** [comment]
Hi Rachel, Mike,

I've looked through most of the code now.  Looks good!  One caveat, I'm not as familiar with the config code, so you should only consider that "lightly" reviewed.

One last question/comment though:

I tried the following experiment to check on the tasks/jobs machinery.  From directory `$GALSIM_DIR/examples/great3/`, I ran (in three different terminals):

```
galsim cgc.yaml output.nfiles=10 output.dir=all10
galsim cgc.yaml output.nfiles=10 output.dir=5by5 -n 2 -j 1
galsim cgc.yaml output.nfiles=10 output.dir=5by5 -n 2 -j 2
```

My expectation was that the results in all10/ would bit-for-bit match those in 5by5/.  What I found was that all10/image-00[0-7]-0.fits matched 5by5/image-00[0-7]-0.fits, and similarly for the epoch_catalog files, but that numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images).  It's as if a random number somewhere got skipped.  Is this surprising behavior?

File: examples/great3/cgc.yaml (hunk 16 above)
Comment (jmeyers314) [inline examples/great3/cgc.yaml]: Great.  Thanks.

**Comment by rmjarvis** [comment]
> My expectation was that the results in all10/ would bit-for-bit match those in 5by5/. What I found was that all10/image-00[0-7]-0.fits matched 5by5/image-00[0-7]-0.fits, and similarly for the epoch_catalog files, but that numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images). It's as if a random number somewhere got skipped. Is this surprising behavior?

Yes, that's probably indicating a bug.  Thanks for checking that.  I'll investigate and see if I can figure out what is going on there.  They should have been identical.

**Comment by rmjarvis** [comment]
> numbers 8 and 9 in the series differed (rather significantly when I ds9'd the images). It's as if a random number somewhere got skipped. Is this surprising behavior?

I figured it out.  The problem was with the cosmos_cat.  The number of objects being rejected was coming out wrong.  I tracked it down to where `psf_fwhm` gets the current value of `psf.items.0.fwhm`.  Querying this when it hadn't already been calculated meant that it had to calculate it then, which should have been fine, but it jumped over the place where we tell it to use `index_key : image_num` (at the top of `psf`).

TL;DR, I fixed it here: 44ca3067cb70ebaa99375b09cafb01dfc2624dc8

**Comment by rmandelb** [comment]
I wonder if there is a short/quick unit test we can add that will be able to catch the kind of issue that Josh found?

**Comment by rmandelb** [comment]
I deprecated ```image_dir``` and ```noise_dir```, and updated the documentation to reflect this, in https://github.com/GalSim-developers/GalSim/commit/56e04d0c6b3b84a0cf493198d36dfd04d5d448c8 . This caused an issue for ```__repr__``` and therefore for pickling, which I fixed in a subsequent commit (I hope...).

@rmjarvis , @jmeyers314 , can you please check this out and see what you think?

**Comment by rmjarvis** [comment]
You don't need to split off the directory in the `__repr__` function.  Providing just a catalog name, complete with the directory, will do that correctly already.  cf. 7c30fdf

**Comment by rmandelb** [comment]
Good point.  Thanks.

**Comment by rmjarvis** [comment]
> I wonder if there is a short/quick unit test we can add that will be able to catch the kind of issue that Josh found?

Good idea.  Done.  	bb87871

**Comment by rmjarvis** [comment]
I think maybe all the comments have been dealt with, except for the suggestion to add template and modules examples to the demos.  Like I said, I think those will be most appropriate in demo12 and/or 13, so I'd like to push that into issue #510.

If that's ok with you (@jmeyers314 and @rmandelb), then I think we're ready to merge this?

**Comment by jmeyers314** [comment]
>I think we're ready to merge this?

Yep.  Looks good to me.  All tests pass.

**Comment by rmandelb** [comment]
+1
